Launcher START_OR_RESTART

davidkohn88 opened the pull request:
Fixes a potential race condition where a scheduler could be started for a given
database, but before it has shut down a create database command is run,
the start action then would not change the state of the worker but it would be waiting
on the wrong vxid, so not see that the extension exists. There is a similar case for
alter extension. In addition, we correct behavior and cause workers not to start
on every alter extension but rather only when there is, in fact a worker, and when
we are upgrading to a version with a scheduler from a version without a scheduler.

cevian commented on sql/pre_install/bgw_startup.sql:
```diff
@@ -0,0 +1,10 @@
+-- We have to create the force restart function here and run it in order to avoid a race condition
```
I think the comment  can be more clear. The race condition is when a create extension start after the scheduler has been started but create extension commits after the scheduler checks for the existence of the extension.

cevian commented on sql/updates/0.11.0--0.12.0.sql:
```diff
@@ -66,3 +66,14 @@ CREATE TABLE IF NOT EXISTS _timescaledb_catalog.installation_metadata (
 
 INSERT INTO _timescaledb_catalog.installation_metadata SELECT 'install_timestamp', to_timestamp(0);
+
+-- We have to create the force restart function here, in the update script for the first version that should include a background worker  
```
the comment here could be improved as above. I would also be explicit that we are explaining why this is not a plain start but rather a start_or_restart.

cevian commented on src/loader/bgw_interface.c:
```diff
@@ -23,6 +23,8 @@ TS_FUNCTION_INFO_V1(ts_bgw_db_workers_stop);
 
 TS_FUNCTION_INFO_V1(ts_bgw_db_workers_restart);
 
+TS_FUNCTION_INFO_V1(ts_bgw_db_workers_force_restart);
```
I think we should be consistent about start_or_restart or force_restart. I prefer the former

cevian commented on src/loader/bgw_launcher.c:
```diff
@@ -531,24 +559,66 @@ message_stop_action(HTAB *db_htab, BgwMessage *message)
+
+/* Start scheduler if it doesn't exist, restart it if it's started, otherwise set vxid and attempt to advance*/
+static AckResult
+message_force_restart_action(HTAB *db_htab, BgwMessage *message, VirtualTransactionId vxid)
```
also consistency between force_restart and start_or_restart

cevian commented on src/loader/bgw_message_queue.h:
```diff
@@ -9,7 +9,8 @@ typedef enum BgwMessageType
 	START,
-	RESTART
+	RESTART,
+	FORCE_RESTART
```
START_OR_RESTART?

JLockerman commented on sql/pre_install/bgw_startup.sql:
```diff
@@ -0,0 +1,7 @@
+AS '@LOADER_PATHNAME@', 'ts_bgw_db_workers_start_or_restart'
+LANGUAGE C VOLATILE;
+
+SELECT _timescaledb_internal.start_or_restart_background_workers();
```
this should have a terminating `\n`?

amytai commented on src/loader/README.md:
```diff
@@ -16,6 +16,44 @@ version 0.8.0 will have timescaledb-0.8.0.so loaded.
+a scheduler for each database, as it cannot figure out whether a 
+scheduler should exist for a given database. The scheduler is 
+responsible for shutting down if it should not exist.
+
```
`the launcher will attempt to start 
a scheduler for each database. The scheduler is 
responsible for shutting down if it should not exist, i.e.-- if the database does not have the TimescaleDB extension installed.`

amytai commented on src/loader/README.md:
```diff
@@ -16,6 +16,44 @@ version 0.8.0 will have timescaledb-0.8.0.so loaded.
+command, databases cannot be dropped if there are any open connections.
+The `stop` message is idempotent. 
+
+`restart`: Used to stop and restart the scheduler if it is running
```
Lead with `may only be used to restart existing schedulers`? Right now it's buried in `if it is running`.

`The launcher will immediately restart the scheduler process, which will in turn wait for the txn that sent the message to finish. After the txn finishes, the scheduler will die if there is no longer an extension (as in the case of a successful `DROP EXTENSION`), or continue normally if there is a new version.`

amytai commented on src/loader/README.md:
```diff
@@ -16,6 +16,44 @@ version 0.8.0 will have timescaledb-0.8.0.so loaded.
+action avoids a possible race condition where a scheduler had been 
+started, but not given a vxid (or was waiting on an incorrect vxid
+so a `start` message might not see the correct version of TimescaleDB.
+A restart will not work, because in many cases the scheduler will not
```
The sentence beginning `A restart will not work...` belongs in a separate section about the specific 0.11.0-->0.12.0 upgrade.

amytai commented on src/loader/README.md:
```diff
@@ -16,6 +16,44 @@ version 0.8.0 will have timescaledb-0.8.0.so loaded.
+start it if it does not exist. This is used when the extension is 
+created or when it is updated from a version without a scheduler 
+to a version with a scheduler (0.12 and up). The start or restart
+action avoids a possible race condition where a scheduler had been 
```
`where a scheduler is in the middle of being started by another code path, for example via the launcher polling mechanism, and will therefore block any `start` message, despite running on a potentially stale version of TimescaleDB.`

davidkohn88 commented:
To address these concerns: 
1) handling events makes the launcher significantly more brittle and we wanted to avoid having to make changes to the launcher as frequently, this was the main reason we decided on this api. 
2) we could make the start action non-idempotent, but it seemed to me to be more useful to the user if it were idempotent and merely made the worker start if it was not around, but wouldn't do anything if it was already there. But, there is the possibility we could eliminate the simple start action (this, also probably limits future flexibility as I can imagine multiple scenarios where we want a more idempotent start action). 
3) idempotent actions currently can't take vxids because of the way we've implemented vxid waiting in the backend that starts up, we can't send them a new vxid to wait on while the first has already been sent. In fact, I might argue that the start action shouldn't wait on vxids at all, in which case it would actually be idempotent. 
4) `start_or_restart` is not an ambiguous action, we actually want to start the scheduler if it doesn't exist, and if it does exist, restart it, and knowing whether it's create database or any other action will not help here. 

All that said, there is a possible other change that would avoid having the new message type which is to simply send both the start and then restart messages in the specific situations where `start_or_restart` would be used. This would be a bit inefficient unless we changed restart to be idempotent if the backend is already waiting on the vxid that we passed in, I think it would also require changes to start, but not horribly invasive ones.  If you would prefer, I can remove the `start_or_restart` code and make that change. I will make the docs changes either way.

davidkohn88 commented:
Actually, thinking about it further, I think doing start and then restart doesn't actually solve the race condition, just makes it less likely to happen...so let's maybe not do that.

davidkohn88 commented on src/loader/README.md:
```diff
@@ -2,28 +2,83 @@
-to the correct TimescaleDB version for the database as soon
-as possible. For example, a database containing TimescaleDB
-version 0.8.0 will have timescaledb-0.8.0.so loaded.
+1) Load the correct versioned library for each database. Multiple databases in
```
not sure?

erimatnor replied:
Well, I don't recognize it in any case. If you want number list, then it should be `1.` AFAIK.

davidkohn88 commented on src/loader/README.md:
```diff
@@ -2,28 +2,83 @@
+
+##Message types sent to the launcher:
+
+`start`: Used to start the scheduler by the user, it is meant to be an idempotent
```
done.

davidkohn88 commented on src/loader/README.md:
```diff
@@ -2,28 +2,83 @@
+same as if it were run once. It is used mainly to reactivate a scheduler that
+the user had stopped.
+
+`stop`: Used to stop the scheduler immediately, it does not wait on a vxid and
```
done

davidkohn88 commented on src/loader/README.md:
```diff
@@ -2,28 +2,83 @@
+## When/which messages are sent:
+
+Server startup: no message sent. the launcher will attempt to start a scheduler
+for each database. It cannot figure out whether a scheduler should exist for a
```
done (though on the line further down).

davidkohn88 commented on src/loader/README.md:
```diff
@@ -2,28 +2,83 @@
+given database because it can only connect to shared catalogs, a. The scheduler
+is responsible for shutting down if it should not exist.
+
+`CREATE DATABASE`: Essentially the same as server startup, the launcher checks
```
done

davidkohn88 commented on src/loader/README.md:
```diff
@@ -2,28 +2,83 @@
+for new databases each time it wakes up and will start schedulers for any that
+it has not seen before.
+
+`CREATE EXTENSION`: the create script sends a `start_or_restart` message, it
```
done

davidkohn88 commented on src/loader/README.md:
```diff
@@ -2,28 +2,83 @@
+it has not seen before.
+
+`CREATE EXTENSION`: the create script sends a `start_or_restart` message, it
+does not use the `start` message because there is a race condition where a
```
done

davidkohn88 commented on src/loader/bgw_launcher.c:
```diff
@@ -599,6 +599,34 @@ message_restart_action(HTAB *db_htab, BgwMessage *message, VirtualTransactionId
+			break;
+		case DISABLED:
+			scheduler_state_trans_disabled_to_enabled(entry);
+
```
done

davidkohn88 commented on src/loader/README.md:
```diff
@@ -2,28 +2,83 @@
+even while they are waiting.  
+
+`start_or_restart`: Used to either stop and restart the scheduler or start it if
+it does not exist, it waits on a vxid but not given a vxid and, much like
```
just a typo, fixed.

davidkohn88 commented on src/loader/README.md:
```diff
@@ -2,28 +2,83 @@
+
+## When/which messages are sent:
+
+Server startup: no message sent. the launcher will attempt to start a scheduler
```
done.

cevian commented:
FWIW I think a lot of complexity here has to do with avoiding starting a scheduler that was previously stopped during an update of the extension. Which is arguable kinda silly since they will be restarted during a database restart anyway --- another words a "stop" is not a permanent operation. If we are willing to let go of that semantic then I think we can simplify some of this code. But otherwise I think there have to be 3 cases -- each handled differently:

1) Create extension - start scheduler whether or not one existed before.
2) Update extension of > 0.12 - restart a scheduler only if it was previously running.
3) Update extension if < 0.12 - start a scheduler whether or not one existed before.

I think the complexity stems from that. There is some additional complexity that results from some races with VXIDs but I think those too become easier if we simplify this logic and just always start_or_restart schedulers on update.

erimatnor commented:
@davidkohn88  I don't understand your statement: "handling events makes the launcher significantly more brittle."

- First, we are already acting on events in the launcher so I don't understand what you mean by this statement and you don't motivate it (i.e., describe how it makes it brittle).'
- Second, my key point was that when events do occur (e.g., due to a create extension or user start/stop), we do not tell the launcher which event occurred but instead

davidkohn88 commented:
superseded by #811
